### PR TITLE
[Static graph] Changes to enable building roslyn

### PR DESCRIPTION
**Pull request by cdmihai**
Things needed to build `roslyn/Compilers.sln` with no changes to the roslyn repo.

### CODE REVIEW

**Comment by cdmihai** [comment]
cc @AndyGerlicher @smera

File: src/Build/Graph/ProjectGraphNode.cs [hunk 1]
```diff
@@ -35,12 +39,34 @@ internal ProjectGraphNode(ProjectInstance projectInstance)
...
+            // First edge wins, in accordance with vanilla msbuild behaviour when multiple msbuild tasks call into the same logical project
+            edges[(this, reference)] = projectReferenceItem;
+        }
 
-        internal void AddReferencingProject(ProjectGraphNode projectGraphNode) => _referencingProjects.Add(projectGraphNode);
+        internal void RemoveReferences(GraphBuilder.GraphEdges edges)
+        {
+            foreach (var reference in _projectReferences)
+            {
+                ErrorUtilities.VerifyThrow(reference._referencingProjects.Contains(this), "references should point the nodes referencing them");
```
Comment (rainersigwald) [inline src/Build/Graph/ProjectGraphNode.cs]:
```suggestion
                ErrorUtilities.VerifyThrow(reference._referencingProjects.Contains(this), "references should point to the nodes referencing them");
```

File: src/Build/Graph/GraphBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,397 @@
...
+                }
+            }
+
+            return propertyDictionary;
+        }
+
+        internal class GraphEdges
+        {
+            private ConcurrentDictionary<(ProjectGraphNode, ProjectGraphNode), ProjectItemInstance> ReferenceItems =
+                new ConcurrentDictionary<(ProjectGraphNode, ProjectGraphNode), ProjectItemInstance>();
```
Comment (rainersigwald) [inline src/Build/Graph/GraphBuilder.cs]: This data type raises some concerns for me. Can you expand on why it's needed?

File: src/Build/Graph/GraphBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,397 @@
...
+            IEnumerable<ProjectGraphEntryPoint> entryPoints,
+            ProjectCollection projectCollection,
+            ProjectGraph.ProjectInstanceFactoryFunc projectInstanceFactory
+            )
+        {
+            var entryPointConfigurationMetadata = AddGraphBuildPropertyToEntryPoints(entryPoints);
+
+            var allParsedProjects = new ConcurrentDictionary<ConfigurationMetadata, ProjectGraphNode>();
+            var tasksInProgress = new ConcurrentDictionary<ConfigurationMetadata, object>();
+            var projectsToEvaluate = new ConcurrentQueue<ConfigurationMetadata>();
```
Comment (livarcocc) [inline src/Build/Graph/GraphBuilder.cs]: Why not just pass the entryPointConfigurationMetadata to the ConcurrentQueue constructor? It is a List and therefore an IEnumerable. If you do that you won't need the foreach below.

File: src/Build/Graph/ProjectGraph.cs [hunk 4]
```diff
@@ -573,6 +530,20 @@ private static ImmutableList<string> ExpandDefaultTargets(ImmutableList<string>
...
+
+                    var expandedTargets = string.IsNullOrEmpty(targetsString)
+                        ? defaultTargets
+                        : ExpressionShredder.SplitSemiColonSeparatedList(targetsString).ToList();
+
+                    targets = targets
+                        .RemoveAt(i)
+                        .InsertRange(i, expandedTargets);
+
+                    i += defaultTargets.Count;
```
Comment (rainersigwald) [inline src/Build/Graph/ProjectGraph.cs]: `defaultTargets` or `expandedTargets`?

File: src/Build/Graph/ProjectInterpretation.cs [hunk 5]
```diff
@@ -115,7 +115,12 @@ internal static ProjectType GetProjectType(ProjectInstance project)
                     : ProjectType.NonMultitargeting;
         }
 
-        public static void PostProcess(ConcurrentDictionary<ConfigurationMetadata, ProjectGraphNode> allNodes)
+        /// <summary>
+        /// To avoid calling nuget at graph construction time, the graph is initially constructed with outer build nodes referencing inner build nodes.
+        /// However, at build time, for non root outer builds, the inner builds are NOT referenced by the outer build, but by the nodes referencing the
+        /// outer build. Change the graph to mimic this behaviour.
```
Comment (rainersigwald) [inline src/Build/Graph/ProjectInterpretation.cs]: I don't understand. It sounds like "it starts outer->inner, but at build time it's actually outer->inner", but that doesn't make sense. Can you rephrase?

File: src/Build/Graph/GraphBuilder.cs [hunk 6]
```diff
@@ -0,0 +1,397 @@
...
+                ProjectNodes = allParsedProjects.Values.ToList();
+            }
+            else
+            {
+                throw new AggregateException(exceptions);
+            }
+
+            IReadOnlyCollection<ProjectGraphNode> GetGraphRoots(IReadOnlyCollection<ProjectGraphNode> entryPointNodes)
+            {
+                var graphRoots = new List<ProjectGraphNode>(entryPointNodes.Count);
```
Comment (livarcocc) [inline src/Build/Graph/GraphBuilder.cs]: Ok to waste memory here if not all entryPoints end up being root?

File: src/Build/Graph/GraphBuilder.cs [hunk 7]
```diff
@@ -0,0 +1,397 @@
...
+                foreach (var entryPoint in entryPoints)
+                {
+                    var globalPropertyDictionary = CreatePropertyDictionary(entryPoint.GlobalProperties);
+
+                    AddGraphBuildGlobalVariable(globalPropertyDictionary);
+
+                    var configurationMetadata = new ConfigurationMetadata(FileUtilities.NormalizePath(entryPoint.ProjectFile), globalPropertyDictionary);
+                    entryPointConfigurationMetadata.Add(configurationMetadata);
+                }
+                return entryPointConfigurationMetadata;
```
Comment (livarcocc) [inline src/Build/Graph/GraphBuilder.cs]: NIT: empty line after closing bracket.

File: src/Build.UnitTests/Graph/ProjectGraph_Tests.cs [hunk 8]
```diff
@@ -769,6 +851,78 @@ public void GetTargetListsFiltersTargetsForOuterAndInnerBuilds()
             }
         }
 
+        [Fact]
+        public void GetTargetListsDoesNotUseTargetsMetadataOnInnerBuildsFromRootOuterBuilds()
```
Comment (rainersigwald) [inline src/Build.UnitTests/Graph/ProjectGraph_Tests.cs]: May I beg you for a diagram of this?

File: src/Build/Graph/GraphBuilder.cs [hunk 9]
```diff
@@ -0,0 +1,397 @@
...
+            allParsedProjects[configurationMetadata] = graphNode;
+            return graphNode;
+        }
+
+        /// <summary>
+        ///     Load a graph with root node at entryProjectFile
+        ///     Maintain a queue of projects to be processed and evaluate projects in parallel
+        ///     Returns false if loading the graph is not successful
+        /// </summary>
+        private bool FindGraphNodes(
```
Comment (livarcocc) [inline src/Build/Graph/GraphBuilder.cs]: Should this be LoadGraphNodes instead? Find seems to imply a search and I would not expect modifications.

File: src/Build.UnitTests/Graph/ProjectGraph_Tests.cs [hunk 10]
```diff
@@ -1239,6 +1419,13 @@ private static void AssertOuterBuildAsNonRoot(ProjectGraphNode outerBuild, Dicti
 
                     innerBuild.ReferencingProjects.ShouldContain(outerBuildReferencer);
                     innerBuild.ReferencingProjects.ShouldNotContain(outerBuild);
+
+                    graph.TestOnly_Edges.TestOnly_HasEdge((outerBuild, innerBuild)).ShouldBeFalse();
```
Comment (rainersigwald) [inline src/Build.UnitTests/Graph/ProjectGraph_Tests.cs]: Wait, why no edge for outer->inner of the same project?

File: src/Build/Graph/ProjectGraphNode.cs [hunk 11]
```diff
@@ -10,6 +11,7 @@ namespace Microsoft.Build.Experimental.Graph
     /// <summary>
     /// Represents the node for a particular project in a project graph.
     /// </summary>
+    [DebuggerDisplay(@"{ToString()}")]
```
Comment (rainersigwald) [inline src/Build/Graph/ProjectGraphNode.cs]: You could do this without exposing `ToString()` as public API by calling the method something else; might be worth doing.

File: src/Build/Graph/GraphBuilder.cs (hunk 6 above)
Comment (cdmihai) [reply src/Build/Graph/GraphBuilder.cs]: Trimmed the excess memory.

File: src/Build/Graph/GraphBuilder.cs (hunk 9 above)
Comment (cdmihai) [reply src/Build/Graph/GraphBuilder.cs]: It finds and loads them, but does not connect them. What do you think about `FindAndLoadGraphNodes`? Seems a bit superfluous.

**Comment by cdmihai** [comment]
> Can you lay out what the problems were when building Roslyn that you've resolved here? That would help place the review.
> 
> The commit `Read ProjectReferenceItem.Targets` is not landing for me clearly. I think I understand the goal but not the implementation.

Some Roslyn projects set the `Targets` metadata on certain `ProjectReference` items. Static graph did not initially take this metadata into account, and therefore it inferred the wrong entry targets for the references which had that metadata set. So then, in addition to the `.default` symbol I also added `.projectReferenceTargetsOrDefault` and updated the target protocol to use it instead of `.default`.

The actual change implies remembering the project reference metadata from graph construction, so it can be read by `ProjectGraph.GetTargetLists` when it infers entry targets.

File: src/Build/Graph/GraphBuilder.cs (hunk 2 above)
Comment (cdmihai) [reply src/Build/Graph/GraphBuilder.cs]: The change requires keeping extra state for each edge (the project reference metadata for each reference). There's two ways to keep it:
1. Enhance the OM by adding a GraphEdge class, and then have ProjectGraphNode reference lists of GraphEdge objects, rather than directly referencing parent/child ProjectGraphNodes
2. Hack it out by introducing a dictionary.

1. implies either changing the public API to add GraphEdge, or making it private by adding a dictionary on each node, which is memory wasteful. Since it's not clear whether anyone would find the extra public API useful, I decided to just go with 2. and keep everything private. If, in the future, somebody needs to access this data, then we can spend more energy in designing some cute looking API.

File: src/Build/Graph/ProjectGraph.cs (hunk 4 above)
Comment (cdmihai) [reply src/Build/Graph/ProjectGraph.cs]: Good catch! Fortunately the defect does not affect correctness.

File: src/Build/Graph/ProjectInterpretation.cs (hunk 5 above)
Comment (cdmihai) [reply src/Build/Graph/ProjectInterpretation.cs]: Updated comment with example.

File: src/Build.UnitTests/Graph/ProjectGraph_Tests.cs (hunk 8 above)
Comment (cdmihai) [reply src/Build.UnitTests/Graph/ProjectGraph_Tests.cs]: image

File: src/Build.UnitTests/Graph/ProjectGraph_Tests.cs (hunk 10 above)
Comment (cdmihai) [reply src/Build.UnitTests/Graph/ProjectGraph_Tests.cs]: These are outer builds that are not roots. For non-root outerbuilds, the project that references the outer build also references the inner builds.

File: src/Build/Graph/GraphBuilder.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (cdmihai) [reply src/Build/Graph/GraphBuilder.cs]: Good question. Moved it inside the method.

**Comment by cdmihai** [comment]
@livarcocc @rainersigwald responded to all feedback
